mm/kfence.c

```diff
@@ -152,28 +158,44 @@ void guarded_free(void *addr)
 	spin_unlock_irqrestore(&kfence_alloc_lock, flags);
 }
 
+static int find_freelist(struct kmem_cache *c)
+{
+	int i;
+	for (i = 0; i < STORED_FREELISTS; i++) {
+		if (this_cpu_read(stored_freelists[i].cache) == c)
+			return i;
+	}
+	return -1;
+}
+
 void *kfence_alloc_and_fix_freelist(struct kmem_cache *s)
 {
 	unsigned long flags;
 	struct kmem_cache_cpu *c = raw_cpu_ptr(s->cpu_slab);
-	struct kmem_cache *stored = this_cpu_read(stored_cache);
+	int fl, num_fl;
+	struct kmem_cache *stored;
 	void *ret = NULL;
 	struct page *page;
-
-	if (stored && (s == stored)) {
-		spin_lock_irqsave(&kfence_lock, flags);
-		stored = this_cpu_read(stored_cache);
-		if (stored && (s == stored)) {
-			ret = guarded_alloc(s->size);
-			c->freelist = this_cpu_read(stored_freelist);
-			this_cpu_write(stored_freelist, NULL);
-			this_cpu_write(stored_cache, NULL);
-			page = virt_to_page(ret);
-		}
-		spin_unlock_irqrestore(&kfence_lock, flags);
-		pr_info("kfence_alloc_and_fix_freelist returns %px\n", ret);
-		return ret;
-	}
+	void *freelist;
+
+	fl = find_freelist(s);
+	if (fl == -1)
```

Comment:
It'd be helpful if you could add some rudimentary comments for non-trivial things.

AFAIK, you're doing double-checked locking here, right?

Reply:
Right. Currently the locking is somewhat messy, I'll need to revisit it.
Guess for now double-checked locking is actually an overkill, since we haven't started measuring performance.

Reply:
Dropped the double-checked locking.